File: Classes/Training.m
/*
	MyFlightbook for iOS - provides native access to MyFlightbook
	pilot's logbook
 Copyright (C) 2014-2023 MyFlightbook, LLC
 
 This program is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.
 
 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.
 
 You should have received a copy of the GNU General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

//
//  Training.m
//  MFBSample
//
//  Created by Eric Berman on 5/30/14.
//
//

#import "Training.h"
#import "HostedWebViewViewController.h"
#import <MyFlightbook-Swift.h>

@interface Training ()

@end

enum _trainingLinks {cidFirst, cidInstructors = cidFirst, cidStudents, cidReqSignatures, cidEndorsements, cid8710, cidModelRollup, cidTimeRollup, cidAchievements, cidMilestoneProgress, cidLast = cidMilestoneProgress};

@implementation Training

- (void)viewDidLoad
{
    [super viewDidLoad];
}

- (void) viewWillAppear:(BOOL)animated
{
    [self.navigationController setToolbarHidden:YES];
    [self.tableView reloadData];
    [super viewWillAppear:animated];
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return 1;
}

- (BOOL) canViewTraining
{
    return mfbApp().isOnLine && mfbApp().userProfile.isValid;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return cidLast - cidFirst + 1;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    static NSString * cellID = @"TrainingCell";
    
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:cellID];
    if (cell == nil)
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:cellID];

    cell.accessoryType = UITableViewCellAccessoryDisclosureIndicator;

    NSInteger cid = indexPath.row;
    switch (cid)
    {
        case cidEndorsements:
            cell.textLabel.text = NSLocalizedString(@"Endorsements", @"Prompt to view/edit endorsements");
            break;
        case cidAchievements:
            cell.textLabel.text = NSLocalizedString(@"Achievements", @"Prompt to view achievements");
            break;
        case cidMilestoneProgress:
            cell.textLabel.text = NSLocalizedString(@"Ratings Progress", @"Prompt to view Ratings Progress");
            break;
        case cidStudents:
            cell.textLabel.text = NSLocalizedString(@"Students", @"Prompt for students");
            break;
        case cidInstructors:
            cell.textLabel.text = NSLocalizedString(@"Instructors", @"Prompt for Instructors");
            break;
        case cidReqSignatures:
            cell.textLabel.text = NSLocalizedString(@"ReqSignatures", @"Prompt for Requesting Signatures");
            break;
        case cid8710:
            cell.textLabel.text = NSLocalizedString(@"8710Form", @"Prompt for 8710 form");
            break;
        case cidModelRollup:
            cell.textLabel.text = NSLocalizedString(@"ModelRollup", @"Prompt for Model Rollup");
            break;
        case cidTimeRollup:
            cell.textLabel.text = NSLocalizedString(@"TimeRollup", @"Prompt for Time Rollup");
            break;
    }
    
    if (!self.canViewTraining) {
        if (@available(iOS 13.0, *)) {
            cell.textLabel.textColor = UIColor.tertiaryLabelColor;
        } else
            cell.textLabel.textColor = UIColor.grayColor;
    }
    [MFBTheme applyThemedImageNamed:@"training.png" toImageView:cell.imageView];
    return cell;
}

#pragma mark - Table view delegate

- (void) pushAuthURL:(NSString *) szDest
{
    if (![self canViewTraining])
    {
        [self showErrorAlertWithMessage:NSLocalizedString(@"TrainingNotAvailable", @"Error message for training if offline or not signed in")];
        return;
    }
    NSString * szURL = [mfbApp().userProfile authRedirForUser:[NSString stringWithFormat:@"@&d=%@&naked=1", szDest]];

	HostedWebViewViewController * vwWeb = [[HostedWebViewViewController alloc] initWithURL:szURL];
	[self.navigationController pushViewController:vwWeb animated:YES];
}

// In a xib-based application, navigation from a table can be handled in -tableView:didSelectRowAtIndexPath:
- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    NSInteger cid = indexPath.row;
    switch (cid)
    {
        case cidEndorsements:
            [self pushAuthURL:@"endorse"];
            break;
        case cidAchievements:
            [self pushAuthURL:@"badges"];
            break;
        case cidMilestoneProgress:
            [self pushAuthURL:@"progress"];
            break;
        case cidInstructors:
            [self pushAuthURL:@"instructorsFixed"];
            break;
        case cidReqSignatures:
            [self pushAuthURL:@"reqSigs"];
            break;
        case cidStudents:
            [self pushAuthURL:@"studentsFixed"];
            break;
        case cid8710:
            [self pushAuthURL:@"8710"];
            break;
        case cidTimeRollup:
            [self pushAuthURL:@"TimeRollup"];
            break;
        case cidModelRollup:
            [self pushAuthURL:@"ModelRollup"];
            break;
    }
}

@end

File: Classes/Training.swift
/*
    MyFlightbook for iOS - provides native access to MyFlightbook
    pilot's logbook
 Copyright (C) 2014-2023 MyFlightbook, LLC
 
 This program is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.
 
 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.
 
 You should have received a copy of the GNU General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

//
//  Training.swift
//  MyFlightbook
//
//  Created by Eric Berman on 3/26/23.
//

import Foundation

@objc public class Training : UITableViewController {
    enum _trainingLinks : Int, CaseIterable {
        case instructors = 0
        case students
        case reqSignatures
        case endorsements
        case form8710
        case modelRollup
        case timeRollup
        case achievements
        case milestoneProgress
    }
    
    public override func viewWillAppear(_ animated: Bool) {
        navigationController?.setToolbarHidden(true, animated: true)
        tableView.reloadData()
        super.viewWillAppear(animated)
    }
    
    func canViewTraining() -> Bool {
        return MFBNetworkManager.shared.isOnLine && MFBProfile.sharedProfile.isValid()
    }
    
    // MARK: Table view data source
    @objc public override func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    
    @objc public override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return _trainingLinks.allCases.count
    }
    
    @objc public override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cellID = "TrainingCell"
        let cell = tableView.dequeueReusableCell(withIdentifier: cellID) ?? UITableViewCell(style: .default, reuseIdentifier: cellID)
        
        var config = cell.defaultContentConfiguration()
        
        cell.accessoryType = .disclosureIndicator
        
        let cid = _trainingLinks(rawValue: indexPath.row)!  // we want to crash if this fails!
        switch cid {
        case .endorsements:
            config.text = String(localized: "Endorsements", comment: "Prompt to view/edit endorsements")
        case .achievements:
            config.text = String(localized: "Achievements", comment: "Prompt to view achievements")
        case .milestoneProgress:
            config.text = String(localized: "Ratings Progress", comment: "Prompt to view Ratings Progress")
        case .students:
            config.text = String(localized: "Students", comment: "Prompt for students")
        case .instructors:
            config.text = String(localized: "Instructors", comment: "Prompt for Instructors")
        case .reqSignatures:
            config.text = String(localized: "ReqSignatures", comment: "Prompt for Requesting Signatures")
        case .form8710:
            config.text = String(localized: "8710Form", comment: "Prompt for 8710 form")
        case .modelRollup:
            config.text = String(localized: "ModelRollup", comment: "Prompt for Model Rollup")
        case .timeRollup:
            config.text = String(localized: "TimeRollup", comment: "Prompt for Time Rollup")
        }
        
        if !canViewTraining() {
            config.textProperties.color = .tertiaryLabel
        }

        MFBTheme.addThemedImageToCellConfig(name: "training.png", config: &config)
        
        cell.contentConfiguration = config

        return cell
    }
    
    // MARK: Table view delegate
    func pushAuthURL(_ szDest : String) {
        if !canViewTraining() {
            showErrorAlertWithMessage(msg: String(localized: "TrainingNotAvailable", comment: "Error message for training if offline or not signed in"))
        } else {
            let szURL = MFBProfile.sharedProfile.authRedirForUser(params: "d=\(szDest)&naked=1")
            let vwWeb = HostedWebViewController(url: szURL)
            navigationController?.pushViewController(vwWeb, animated: true)
        }
    }
    
    // In a xib-based application, navigation from a table can be handled in -tableView:didSelectRowAtIndexPath:
    public override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        
        let cid = _trainingLinks(rawValue: indexPath.row)!
        switch (cid)
        {
        case .endorsements:
            pushAuthURL("endorse")
        case .achievements:
            pushAuthURL("badges")
        case .milestoneProgress:
            pushAuthURL("progress")
        case .instructors:
            pushAuthURL("instructorsFixed")
        case .reqSignatures:
            pushAuthURL("reqSigs")
        case .students:
            pushAuthURL("studentsFixed")
        case .form8710:
            pushAuthURL("8710")
        case .timeRollup:
            pushAuthURL("TimeRollup")
        case .modelRollup:
            pushAuthURL("ModelRollup")
        }
    }
}
